### ISSUE TITLE
Rewrite LDAP backend

### ISSUE THREAD

**Issue by jefferai** [hashicorp/vault#546]
The LDAP backend tries to hide complexity from the user; instead it increases complexity by requiring an increasing amount of logic and role parameters to handle various LDAP setups.

The backend should follow the example of other LDAP-integrating programs and allow for flexible and expressive parameters that put the user in control. This will also simplify the internal logic quite a bit.

**Comment by hobbeswalsh** [comment]
I'm sure my specific use-case is covered by the general and thoughtful comment at the top of this issue, but I'd just like to :+1: the capability to specify a bind user and password.

**Comment by jefferai** [comment]
This is unfortunately slipping to 0.4, not because it's not a high priority (it is!) but because other items needed fixing first, and time has simply run out for 0.3. As this is worked on, I'll keep this ticket updated.

**Comment by jhaals** [comment]
I'm ok with waiting.
It might be good to formalise the layout of the updated backend as well since every ldap setup seem to be somewhat unique.

**Comment by hobbeswalsh** [comment]
I'd love to see it formalized because I'd like to potentially contribute if
possible. Thanks!


On Mon, Sep 21, 2015 at 07:18 Johan Haals <notifications@github.com> wrote:

> I'm ok with waiting.
> It might be good to formalise the layout of the updated backend as well
> since every ldap setup seem to be somewhat unique.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hashicorp/vault/issues/546#issuecomment-141996163>.
>
-- 
Brevity: 👻 of 💡 (sent from a phone)

**Comment by jefferai** [comment]
Do you mean the API pathing/routing?

**Comment by hobbeswalsh** [comment]
Sure, if that's what serves as a spec for the auth backend, that would be
excellent.
On Mon, Sep 21, 2015 at 07:23 Jeff Mitchell <notifications@github.com>
wrote:

> Do you mean the API pathing/routing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hashicorp/vault/issues/546#issuecomment-141997449>.
>
-- 
Brevity: 👻 of 💡 (sent from a phone)

**Comment by jhaals** [comment]
Yeah and "features" that should be added to the backend, some asked for a specific bind user and I have issues with the backend in it's current form due to our user DNs being `cn=Firstname Lastname,ou=users`, that does not match on posixGroups where the `memberUid:` maps to the users `uid`.

I solved it with this patch for now which searches for the username before doing the bind to get around the problem with us having cn= instead of uid= in our dn
```
// Find dn for username
sresult, err := c.Search(&ldap.SearchRequest{
     BaseDN: cfg.UserDN,
     Scope:  2, // subtree
     Filter: fmt.Sprintf("(uid=%s)", username),
})
if err != nil {
     return nil, logical.ErrorResponse(fmt.Sprintf("LDAP bind failed: %v", err)), nil
}
for _, e := range sresult.Entries {
     binddn = e.DN
}
```

**Comment by jefferai** [comment]
I would be happy to discuss a spec/API design document with anyone interested. I won't have the time to work on this during this week or next, so if anyone wants to start one up, feel free.

My thoughts:

* LDAP ain't easy, so don't try to hide all the gory details. For any possible LDAP bind/auth/search capability X, figure that admins know what they're doing and may well need to toggle X.
* Most LDAP parameters are specified in bind URLs or DNs; I'm more than happy to allow operators to specify those strings verbatim (with placeholders for necessary substitution) than try to have a jazillion options used to construct those.
* I've used LDAP with a number of various software packages; I'm planning on going through the settings page of those and seeing what the common (and uncommon) parameters are. Anything common that isn't handled is a failure. Anything uncommon is a bonus, but if it's super uncommon and will be terrible to accommodate, don't do so unless people are asking for it.

**Comment by Moeser** [comment]
Just signed up to github so I can comment on this.  I wanted to mention our use case so it can be thought about during the re-write.  We have multiple userdn's, (one for users, one for service accounts, etc.)  If we can only specify one userdn then a subset of users is excluded.

Please also consider support for multiple userdn's :)

**Comment by Poohblah** [comment]
@Moeser - If I'm understanding you correctly, in a typical LDAP client, you don't need to specify multiple DNs to achieve your goal.  Instead, you specify a base DN (typically an OU) under which all user accounts live (may even be the top-level entity in your org) and an ObjectClass that is common to all user accounts. For instance, `userSearchBase = ou=people,dc=example,dc=org` and `userObjectClass = person`. This configuration, if an LDAP client is implemented correctly, should catch all entities in, say, `ou=serviceAccounts,ou=people,dc=example,dc=org` as well. More complex scenarios can be handled by an arbitrary search filter.

In other words, I (personally) haven't seen support for multiple base user DNs in the wild, and since you can work around this with other typical configuration options instead, it isn't necessary for an LDAP client to support multiple base user DNs.

**Comment by jefferai** [comment]
Also, nearly all backends in Vault can be mounted in more than one
place. So you could have multiple backends with multiple
configurations. We may also follow the role-based model of many of the
secret backends, so you can define multiple roles with different
characteristics and have users auth to one or the other. (Plus, even
with roles you can mount in multiple places!)

**Comment by swill** [comment]
I am also keenly interested in this as I have a new application in development which will require this functionality.  

I am currently using the following setup:
```
$ vault write auth/ldap/config url="ldaps://my_endpoint:636" upndomain="my.adfs.com" userattr=sAMAccountName userdn="CN=Users,DC=my,DC=adfs,DC=com" groupdn="CN=Domain Users,CN=Users,DC=my,DC=adfs,DC=com" insecure_tls=true starttls=true

$ vault write auth/ldap/groups/"Domain Users" policies=root

$ vault auth -method=ldap username=myuser
Password (will be hidden): 
Error making API request.

URL: PUT http://127.0.0.1:8200/v1/auth/ldap/login/myuser
Code: 400. Errors:

* user is not member of any authorized group
```

I see others having the same issue here #596.

If I can help, let me know.  I am a professional developer and I plan to develop my current project with a dependancy on this functionality...

**Comment by jefferai** [comment]
An update for all of you:

We've been doing a lot of roadmap setting internally and one of the things that we're planning on for the 0.5/0.6 timeframe is a major rewrite of the frameworks that backends are built upon, to make them far more standardized, understandable, better able to support needed features (like listing), and so on. Along with this will come major rewrites of nearly all backends.

So the bad news is that there is currently no progress on this yet and we're looking to do 0.4 in the near term. The good news is that LDAP will be one of the first backends rewritten, used to help define and test the new frameworks.

One of the other items being lined up is enhancements to the test framework that will let us spin up Docker containers with daemons and data for testing. What this means is that anyone interested can help provide the test cases used in the new backend's acceptance tests. This would require information about which Docker container to use and the test data necessary to be loaded into the LDAP daemon inside to then run queries against. I realize that this doesn't help AD users...one possible workaround would be to enable Vagrant boxes as an alternative, and another would be for interested users to come up with OpenLDAP or other LDAP daemon schemas that, in their testing, behaves identically to their AD setups.

**Comment by mtougeron** [comment]
Is there any update on this ticket?

**Comment by freimer** [comment]
You don't need a bind DN.  This, much like Elasticsearch, uses the user and their password themselves to bind to LDAP.  If the user DN is wrong, or the password is wrong, the bind will fail.  Please don't modify this to require a special bind DN.  We have enough "service" accounts already, and not requiring one for this is great.  Your problems are likely incorrect userattr.  I had to change mine to CN, and use the correct userdn, to get it to work.  sAMAccountName did NOT work.  CN did.  Try this.

**Comment by jefferai** [comment]
@freimer You don't need a bind DN but some do. You don't need sAMAccountName but some do. Nobody is talking about requiring either of these things but it does need to be supported.

**Comment by jefferai** [comment]
@mtougeron still planned, no other update at the moment... sorry.

**Comment by freimer** [comment]
@jefferai Cool.  I don't have any problem if you *add* functionality.  Just suggesting some options that may help others.

**Comment by mtougeron** [comment]
@jefferai thanks for the update

**Comment by Moeser** [comment]
Any status updates here?  Comments indicate that the release target was adjusted to a 0.5.0 or 0.6.0 release.  0.5.2 is already out, and I think the next release is 0.6.0.  Did this work fall off of the roadmap?

**Comment by jefferai** [comment]
Currently we've had a number of pull requests that have implemented needed features, so the urgency for this has fallen off. Also, the rework of the backend framework that this would be rebased on has also been pushed back indefinitely. 

At the moment, if you're having specific problems with the current backend we can address those issues individually.

**Comment by jefferai** [comment]
I'm going to close this ticket, because with the many pull requests that have come in for 0.5.x and 0.6.0 the backend is already mostly rewritten. Further changes and enhancements are set to come in via #1388 . If you have specific problems or requests, please open issues for them. Thanks!
